src/OmniSharp.Roslyn.CSharp/Services/IntelliSense/V2/CompletionService.cs

```diff
@@ -72,11 +76,114 @@ public CompletionService(OmniSharpWorkspace workspace, ILoggerFactory loggerFact
                 };
             }
 
-            return new Models.CompletionResponse
+            var response = new CompletionModels.CompletionResponse
             {
                 IsSuggestionMode = isSuggestionMode,
                 Items = items
             };
+
+            _lastResponse = (fileName, completionList);
+
+            return response;
+        }
+
+        private static CSharpCompletionService GetService(Document document)
+        {
+            var service = CSharpCompletionService.GetService(document);
+            if (service == null)
+            {
+                throw new InvalidOperationException("Could not retrieve Roslyn CompletionService.");
+            }
+
+            return service;
+        }
+
+        private Document GetDocument(string fileName)
+        {
+            var document = _workspace.GetDocument(fileName);
+            if (document == null)
+            {
+                throw new ArgumentException($"Could not find document for {fileName}.");
+            }
+
+            return document;
+        }
+
+        public async Task<CompletionModels.CompletionItemResolveResponse> Handle(CompletionModels.CompletionItemResolveRequest request)
+        {
+            if (_lastResponse.fileName == null || _lastResponse.completionList == null)
+            {
+                throw new InvalidOperationException($"{OmniSharpEndpoints.V2.CompletionItemResolve} end point cannot be called before {OmniSharpEndpoints.V2.Completion}");
+            }
+
+            var previousFileName = _lastResponse.fileName;
+            var previousCompletionList = _lastResponse.completionList;
+
+            var fileName = request.FileName;
+            var itemIndex = request.ItemIndex;
+            var displayText = request.DisplayText;
+
+            if (!string.Equals(_lastResponse.fileName, fileName, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new ArgumentException($"Cannot resolve completion item from '{fileName}' because the last {OmniSharpEndpoints.V2.Completion} request was for '{previousFileName}'");
+            }
+
+            if (itemIndex < 0 || itemIndex >= previousCompletionList.Items.Length)
+            {
+                throw new ArgumentOutOfRangeException($"Invalid item index: {itemIndex}. Should be within range 0 to {previousCompletionList.Items.Length}");
+            }
+
+            var previousItem = previousCompletionList.Items[itemIndex];
+            if (!string.Equals(previousItem.DisplayText, displayText))
```

Comment:
The doc comment makes it sound like setting `displayText` is optional, but this will always fail